wordpress.org/public_html/wp-content/plugins/wporg-profiles-wp-activity-notifier/wporg-profiles-wp-activity-notifier.php

```diff
@@ -150,6 +151,71 @@ public function maybe_notify_new_published_post( $new_status, $old_status, $post
 		$this->notify_new_blog_post( $post );
 	}
 
+	public function insert_post ($postId, $post) {
+		if ( defined( 'REST_REQUEST' ) && REST_REQUEST ) {
```

Comment:
Avoid the second action firing from Gutenberg.

Reply:
Is `REST_REQUEST` a reliable way to detect that sequence of events? And is that a common occurrence?

At a quick glance it seems some of the other notification types don't try to prevent duplicate activities when something is published twice. It might be overkill, but if it's common then as a general solution I would consider using postmeta to store a list of notified actions to prevent dupes. Perhaps that's better done on the profile API side - @dd32 do you have an opinion on this?

Reply:
Since `wp_insert_post` would be fired twice because of Gutenberg (according to these two posts: [post1](https://wordpress.stackexchange.com/questions/395666/why-is-save-post-hook-being-called-twice-despite-all-my-efforts), [post2](https://wordpress.org/support/topic/publish_post-hook-trigger-twice-when-i-publish-post/)) and caused the issue showed in the screencast below. (You can see that there is another updating... after published, making the account rentest have duplicate activities for the assigned presenter)

https://user-images.githubusercontent.com/18050944/178351308-647ccfc7-e6eb-460e-89c1-e7e7b6f9efa9.mp4

I first tried the suggested solution brought up in post1, which is to use [wp_after_insert_post](https://developer.wordpress.org/reference/functions/wp_after_insert_post/) instead, but the result didn't change.

So I then use the method mentioned in post2, which is using `REST_REQUEST` to detect the REST API fired by the Gutenberg editor.

> using postmeta to store a list of notified actions to prevent dupes

I might need to ask you more about this.

Reply:
> Since wp_insert_post would be fired twice because of Gutenberg

You should be able to look at the metadata of the insert I believe, and see the state change somehow (I'm guessing post_status: auto-draft -> draft -> published maybe?)

Reply:
> It might be overkill, but if it's common then as a general solution I would consider using postmeta to store a list of notified actions to prevent dupes.

Dupes should be able to be handled on the profiles side, every activity entry has a contextual link/id/etc.. but avoiding sending duplicate data in the same pageload is obviously beneficial from a performance standpoint..

Reply:
Hm, that's interesting about the extra call to wp_insert_post. I wonder if this is the cause:

https://wordpress.stackexchange.com/questions/395666/why-is-save-post-hook-being-called-twice-despite-all-my-efforts#comment576840_395672

> Some plugins incorrectly make a second REST API request to save meta after a post has been saved because the author doesn't understand the core data API and doesn't realise Gutenberg will send the meta for them if they register it and update the posts meta directly. This is where that comes from, and it's bad practice

I don't know how the presenter metadata is handled, but it's possible that (or other postmeta) is causing the duplicate call. That might explain why you were having trouble fetching the postmeta from the `transition_post_status` hook - maybe the meta is only stored in the second call, after the post status has changed. I'll do some testing and see what I can turn up.

Reply:
I found some discussion here, I guess it explains why this happened.
https://github.com/WordPress/gutenberg/issues/12903#issue-391380038
> As we know that Gutenberg sends two requests when we hit the Publish/Update button:
a) Request to WP REST API to create/update the post
b) Request to post.php to save/submit the metaboxes

> a) Gutenberg saves data via one or two requests; the second save request only fires if there are meta boxes, so a plugin cannot depend on this request happening.
b) If there are going to be two requests, the plugin can't necessarily depend on the first request, because the first doesn't contain all the data for the post.